include/nbl/builtin/hlsl/cpp_compat/intrinsics.hlsl

```diff
@@ -0,0 +1,302 @@
+#include <nbl/builtin/hlsl/spirv_intrinsics/GLSL.std.450.hlsl>
+
+#ifndef __HLSL_VERSION
+#include <algorithm>
+#include <cmath>
+#include "nbl/core/util/bitflag.h"
+#endif
+
+namespace nbl
+{
+namespace hlsl
+{
+
+template<typename Integer>
+int bitCount(NBL_CONST_REF_ARG(Integer) val)
+{
+#ifdef __HLSL_VERSION
+	return countbits(val);
+#else
+	return glm::bitCount(val);
+#endif
+}
+
+template<typename T>
+vector<T, 3> cross(NBL_CONST_REF_ARG(vector<T, 3>) lhs, NBL_CONST_REF_ARG(vector<T, 3>) rhs)
+{
+#ifdef __HLSL_VERSION
+	return spirv::cross(lhs, rhs);
+#else
+	return glm::cross(lhs, rhs);
+#endif
+}
+
+template<typename T>
+T clamp(NBL_CONST_REF_ARG(T) val, NBL_CONST_REF_ARG(T) min, NBL_CONST_REF_ARG(T) max)
+{
+#ifdef __HLSL_VERSION
+	return clamp(val, min, max);
+#else
+	return glm::clamp(val, min, max);
+#endif
+}
+
+namespace dot_product_impl
+{
+template<typename T>
+struct dot_helper
+{
+	using scalar_type = typename vector_traits<T>::scalar_type;
+
+	static inline scalar_type dot(NBL_CONST_REF_ARG(T) lhs, NBL_CONST_REF_ARG(T) rhs)
+	{
+		static array_get<T, scalar_type> getter;
+		scalar_type retval = getter(lhs, 0) * getter(rhs, 0);
+
+		static const uint32_t ArrayDim = sizeof(T) / sizeof(scalar_type);
+		for (uint32_t i = 1; i < ArrayDim; ++i)
+			retval = retval + getter(lhs, i) * getter(rhs, i);
+
+		return retval;
+	}
+};
+
+#define DEFINE_BUILTIN_VECTOR_SPECIALIZATION(FLOAT_TYPE, RETURN_VALUE)\
+template<uint32_t N>\
+struct dot_helper<vector<FLOAT_TYPE, N> >\
+{\
+	using VectorType = vector<FLOAT_TYPE, N>;\
+	using ScalarType = typename vector_traits<VectorType>::scalar_type;\
+\
+	static inline ScalarType dot(NBL_CONST_REF_ARG(VectorType) lhs, NBL_CONST_REF_ARG(VectorType) rhs)\
+	{\
+		return RETURN_VALUE;\
+	}\
+};\
+
+#ifdef __HLSL_VERSION
+#define BUILTIN_VECTOR_SPECIALIZATION_RET_VAL dot(lhs, rhs)
+#else
+#define BUILTIN_VECTOR_SPECIALIZATION_RET_VAL glm::dot(lhs, rhs)
+#endif
+
+DEFINE_BUILTIN_VECTOR_SPECIALIZATION(float16_t, BUILTIN_VECTOR_SPECIALIZATION_RET_VAL)
+DEFINE_BUILTIN_VECTOR_SPECIALIZATION(float32_t, BUILTIN_VECTOR_SPECIALIZATION_RET_VAL)
+DEFINE_BUILTIN_VECTOR_SPECIALIZATION(float64_t, BUILTIN_VECTOR_SPECIALIZATION_RET_VAL)
+
+#undef BUILTIN_VECTOR_SPECIALIZATION_RET_VAL
+#undef DEFINE_BUILTIN_VECTOR_SPECIALIZATION
+
+}
+
+template<typename T>
+typename vector_traits<T>::scalar_type dot(NBL_CONST_REF_ARG(T) lhs, NBL_CONST_REF_ARG(T) rhs)
+{
+	return dot_product_impl::dot_helper<T>::dot(lhs, rhs);
+}
+
+// determinant not defined cause its implemented via hidden friend
+// https://stackoverflow.com/questions/67459950/why-is-a-friend-function-not-treated-as-a-member-of-a-namespace-of-a-class-it-wa
+template<typename T, uint16_t N, uint16_t M>
+inline T determinant(NBL_CONST_REF_ARG(matrix<T, N, M>) m)
+{
+#ifdef __HLSL_VERSION
+
+#else
+	return glm::determinant(reinterpret_cast<typename matrix<T, N, M>::Base const&>(m));
+#endif
+}
+
+template<typename Integer>
+int findLSB(NBL_CONST_REF_ARG(Integer) val)
```

Comment:
clamp, floor, and everything below this line should have overloads for vectors.

Some of these like clamp, min, max, floor, etc. actually work on vectors too in SPIR-V

Reply:
no need for overloads, it will work